Chat_App/Chat_App/UI/TypingComponentView.swift

```diff
@@ -0,0 +1,188 @@
+    private func setUpSubviews() {
+        addSubview(containerView)
+        containerView.addSubview(button)
+        containerView.addSubview(textView)
+    }
+
+    func setUpTextView(with color: UIColor) {
+        textView.textColor = color
+    }
+
+
+
+
+    // MARK: - Layout Constraints
+    private func setUpLayoutConstraints() {
+        setUpContainerViewLayoutConstraints()
+        setUpTextViewLayoutConstraints()
+        setUpButtonLayoutConstraints()
+    }
+
+    private func setUpContainerViewLayoutConstraints() {
+        NSLayoutConstraint.activate([
+            containerView.topAnchor.constraint(equalTo: topAnchor),
+            containerView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: Constants.containerLeadingPadding),
+            containerView.bottomAnchor.constraint(equalTo: bottomAnchor),
+            containerView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: Constants.containerTrailingPadding),
+            containerView.heightAnchor.constraint(lessThanOrEqualToConstant: Constants.containerViewMaxHeigth)
+        ])
+    }
+
+    private func setUpTextViewLayoutConstraints() {
+        NSLayoutConstraint.activate([
+            textView.topAnchor.constraint(equalTo: containerView.topAnchor, constant: Constants.textViewtopLeadingPadding),
+            textView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor, constant: Constants.textViewbottomPadding),
+            textView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor, constant: Constants.textViewtopLeadingPadding),
+            textView.trailingAnchor.constraint(equalTo: button.leadingAnchor, constant: Constants.textViewtrailingPadding)
+        ])
+    }
+
+    private func setUpButtonLayoutConstraints() {
+        NSLayoutConstraint.activate([
+            button.bottomAnchor.constraint(equalTo: containerView.bottomAnchor, constant: Constants.sendButtonBottomPadding),
+            button.trailingAnchor.constraint(equalTo: containerView.trailingAnchor, constant: Constants.containerTrailingPadding),
+            button.heightAnchor.constraint(equalToConstant: Constants.sendButtonWidthAndHeight),
+            button.widthAnchor.constraint(equalToConstant: Constants.sendButtonWidthAndHeight)
+        ])
+    }
+
+    // MARK: Actions
+    private func calculateMaxLines() -> Int {
+        let maxSize = CGSize(width: textView.frame.width, height: CGFloat(Float.infinity))
+        let font = textView.font ?? .systemFont(ofSize: Constants.textViewFontSize)
+        let charSize = font.lineHeight
+        let text = (textView.text ?? "") as NSString
+        let textSize = text.boundingRect(with: maxSize, options: .usesLineFragmentOrigin, attributes: [NSAttributedString.Key.font: font], context: nil)
+        let linesRoundedUp = Int(ceil(textSize.height/charSize))
+        return linesRoundedUp
+    }
+
+    private func updateTextViewHeight() {
+        if calculateMaxLines() > Constants.maxLineNubmer {
```

Comment:
```suggestion
        let value = calculateMaxLines() > Constants.maxLineNubmer
        heightConstraint?.isActive = value
        textView.isScrollEnabled = value
```
would be better?

Reply:
private func updateTextViewHeight() {
        let value = calculateMaxLines() > Constants.maxLineNubmer
        if value {
            if heightConstraint != nil { return }
            heightConstraint = containerView.heightAnchor.constraint(equalToConstant: containerView.frame.height)
            heightConstraint?.isActive = true
            textView.isScrollEnabled = true
        } else {
            textView.isScrollEnabled = false

            if let heightConstraint = heightConstraint {
                heightConstraint.isActive = false
                containerView.removeConstraint(heightConstraint)
                containerView.layoutIfNeeded()
                textView.layoutIfNeeded()
                layoutIfNeeded()
            }
            heightConstraint = nil
        }
    }

i change it like this , this one is more readability